### DISCUSSION TITLE
Approach for tagging grammatical patterns (Japanese)

### DISCUSSION THREAD

**Discussion by e-e** [explosion/spaCy#10044] [Help: Other Questions] [answered]
I'm looking for some direction regarding the best approach to defining and tagging grammatical patterns in Japanese. I have taken a bit of a break since the last time I was toying with spaCy for this task, so I'm a bit rusty, but I know previously I was wondering if Prodigy would be a good approach for this. I also attempted manually constructing matchers, but wasn't totally able to figure out some edge cases (very likely a result of not being super familiar with the library though). Anyways, any help regarding an approach would be very much appreciated - thank you in advance!

**Comment by polm** [comment, marked as answer]
What is your actual question / goal? I don't remember what you are referring to. The DependencyMatcher should be able to match many types of grammatical patterns.

**Reply by e-e** [reply]
The end goal is that I would like to tag sections of text and map those to "grammar points" as they are taught in different textbooks

**Reply by e-e** [reply]
For example, I build this PoC a while back with Mecab and RoR, and it was a huge pain to create all the matching rules manually, but this is the basic idea I'm going for:

<img width="354" alt="Screen Shot 2022-01-13 at 2 18 06 PM" src="https://user-images.githubusercontent.com/5191838/149403129-bfe3eeb5-5f18-4b2b-b4d0-af3d52bd0f29.png">

<img width="399" alt="Screen Shot 2022-01-13 at 2 18 14 PM" src="https://user-images.githubusercontent.com/5191838/149403150-aa98d8b1-57c0-4204-98d3-5ce40ad1c1a9.png">

<img width="371" alt="Screen Shot 2022-01-13 at 2 18 18 PM" src="https://user-images.githubusercontent.com/5191838/149403167-512edc9c-d344-4180-a45f-274ece37b598.png">

**Reply by e-e** [reply]
Essentially I want to tag each grammar point with some unique id which is associated to the grammar point in a database.

**Reply by polm** [reply]
Since any textbook's list of "grammar points" is to some degree arbitrary, I don't think there's any way you can do better than manually writing rules for them, though it looks like the actual rules should be pretty simple. 

For IDs, if you make a list of IDs for each grammar point, you should be able to use the [EntityRuler ID feature](https://spacy.io/usage/rule-based-matching#entityruler-ent-ids) to map matches back to grammar points.

**Reply by e-e** [reply]
Would prodigy be useful for this?

**Reply by polm** [reply]
Prodigy would be useful for reviewing your patterns and checking the results. I guess it depends on what you consider a "grammar point", but I would think a rule-based approach would capture them well and it'd be hard to use a statistical model, so you wouldn't benefit much from training data / a model.

**Reply by e-e** [reply]
I'll give the rule-based matching a shot. Thanks for the help!
